### ISSUE TITLE
Weird behavior from torch.stack

### ISSUE THREAD

**Issue by rohit-gupta** [pytorch/pytorch#29510]
## 🐛 Bug

torch.stack results in `RuntimeError: invalid argument 0: Sizes of tensors must match except in dimension 0. Got 1 and 4 in dimension 1` errors for perfectly valid input. At first I thought I was misunderstanding how indexing works, but when I changed "dim" in the stack call to 1, I got the error with `RuntimeError: invalid argument 0: Sizes of tensors must match except in dimension 1. Got 1 and 4 in dimension 0` which is really really weird.

## To Reproduce

1. Run this code snippet:

```python
import torch


class_probs = torch.zeros([4,1024,1024], device="cpu")
bg_probs = torch.zeros([1,1024,1024], device="cpu")

# class_probs = torch.zeros([4,1024,1024], device="cuda")
# bg_probs = torch.zeros([1,1024,1024], device="cuda")

print(bg_probs.shape, class_probs.shape)
combo = torch.stack((bg_probs, class_probs), dim=0)
print(combo.shape)
```

## Environment

PyTorch version: 1.3.0
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 418.87.01
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] torch==1.3.0
[pip] torchcontrib==0.0.2
[pip] torchfile==0.1.0
[pip] torchnet==0.0.4
[pip] torchvision==0.4.1a0+d94043a
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.14           py37ha843d7b_0  
[conda] mkl_random                1.1.0            py37hd6b4f25_0  
[conda] pytorch                   1.3.0           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchcontrib              0.0.2                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchnet                  0.0.4                    pypi_0    pypi
[conda] torchvision               0.4.1                py37_cu101    pytorch

## Additional context

The error happens for both cuda and cpu tensors. Also, torch.cat seems to work perfectly fine for the same operation !

**Comment by SsnL** [comment]
the input is not valid. `stack` adds a new dimension so all dimensions must match exactly. if you want to concatenate along the first dimension, you need `torch.cat` instead.

**Comment by rohit-gupta** [comment]
@SsnL Thanks for the explanation, but the error message is still weird. If all dimensions must match, then the function shouldn't print "Sizes of tensors must match except in dimension 0" as the error message.

**Comment by SsnL** [comment]
Yeah, I agree. I reopened this as a request to improve the error message. The cause is that `stack` is implemented as `unsqueeze` + `cat` and the error message is thrown from shape checks in `cat`.
